File: app/modules/planning/domain/calculations/opportunities/rebalance_buys.py
"""Rebalance buy opportunity calculator.

Identifies underweight areas that should be increased for rebalancing.
"""

from typing import Any, Dict, List, Optional

from app.domain.services.exchange_rate_service import ExchangeRateService
from app.domain.value_objects.trade_side import TradeSide
from app.modules.planning.domain.calculations.opportunities.base import (
    OpportunityCalculator,
    OpportunityContext,
    opportunity_calculator_registry,
)
from app.modules.planning.domain.models import ActionCandidate
from app.modules.trading.domain.trade_sizing_service import TradeSizingService


class RebalanceBuysCalculator(OpportunityCalculator):
    """Identifies underweight areas to increase for rebalancing."""

    def __init__(self, exchange_rate_service: Optional[ExchangeRateService] = None):
        self.exchange_rate_service = exchange_rate_service

    @property
    def name(self) -> str:
        return "rebalance_buys"

    def default_params(self) -> Dict[str, Any]:
        return {
            "underweight_threshold": 0.05,  # 5%+ underweight triggers rebalance
            "base_trade_amount_eur": 1000.0,  # Base trade amount
        }

    async def calculate(
        self,
        context: OpportunityContext,
        params: Dict[str, Any],
    ) -> List[ActionCandidate]:
        """
        Identify rebalance buy opportunities (underweight areas to increase).

        Checks country/industry allocations against targets and suggests
        buys in underweight groups.

        Args:
            context: Portfolio context with allocations and targets
            params: Calculator parameters

        Returns:
            List of buy ActionCandidates for underweight areas
        """
        opportunities: List[ActionCandidate] = []

        # Require allocation data
        if not context.country_allocations or not context.country_to_group:
            return opportunities
        if not context.country_weights:
            return opportunities

        underweight_threshold = params.get("underweight_threshold", 0.05)
        base_trade_amount = params.get("base_trade_amount_eur", 1000.0)

        for security in context.securities:
            if not security.allow_buy:
                continue

            # Get current price from position if exists
            price = None
            pos = next(
                (p for p in context.positions if p.symbol == security.symbol), None
            )
            if pos:
                price = pos.current_price

            if not price or price <= 0:
                continue

            # Get quality score if available
            quality_score = 0.5
            if context.security_scores:
                quality_score = context.security_scores.get(security.symbol, 0.5)

            # Check if this security's country group is underweight
            country = security.country
            if not country:
                continue

            group = context.country_to_group.get(country, "OTHER")
            target = context.country_weights.get(group, 0)
            current = context.country_allocations.get(group, 0)

            if current < target - underweight_threshold:
                underweight = target - current

                # Calculate buy quantity
                exchange_rate = 1.0
                if security.currency and security.currency != "EUR":
                    if self.exchange_rate_service:
                        exchange_rate = await self.exchange_rate_service.get_rate(
                            security.currency, "EUR"
                        )

                sized = TradeSizingService.calculate_buy_quantity(
                    target_value_eur=base_trade_amount,
                    price=price,
                    min_lot=security.min_lot,
                    exchange_rate=exchange_rate,
                )

                # Apply priority multiplier
                base_priority = underweight * 2 + quality_score * 0.5
                security_multiplier = security.priority_multiplier if security else 1.0
                final_priority = base_priority * security_multiplier

                opportunities.append(
                    ActionCandidate(
                        side=TradeSide.BUY,
                        symbol=security.symbol,
                        name=security.name,
                        quantity=sized.quantity,
                        price=price,
                        value_eur=sized.value_eur,
                        currency=security.currency or "EUR",
                        priority=final_priority,
                        reason=f"Underweight {group} by {underweight*100:.1f}%",
                        tags=["rebalance", f"underweight_{group.lower()}"],
                    )
                )

        return opportunities


# Auto-register this calculator
_rebalance_buys_calculator = RebalanceBuysCalculator()
opportunity_calculator_registry.register(
    _rebalance_buys_calculator.name, _rebalance_buys_calculator
)

File: trader-go/internal/modules/opportunities/calculators/rebalance_buys.go
package calculators

import (
	"fmt"
	"sort"

	"github.com/aristath/arduino-trader/internal/modules/planning/domain"
	"github.com/rs/zerolog"
)

// RebalanceBuysCalculator identifies underweight positions to buy for rebalancing.
type RebalanceBuysCalculator struct {
	*BaseCalculator
}

// NewRebalanceBuysCalculator creates a new rebalance buys calculator.
func NewRebalanceBuysCalculator(log zerolog.Logger) *RebalanceBuysCalculator {
	return &RebalanceBuysCalculator{
		BaseCalculator: NewBaseCalculator(log, "rebalance_buys"),
	}
}

// Name returns the calculator name.
func (c *RebalanceBuysCalculator) Name() string {
	return "rebalance_buys"
}

// Category returns the opportunity category.
func (c *RebalanceBuysCalculator) Category() domain.OpportunityCategory {
	return domain.OpportunityCategoryRebalanceBuys
}

// Calculate identifies rebalance buy opportunities.
func (c *RebalanceBuysCalculator) Calculate(
	ctx *domain.OpportunityContext,
	params map[string]interface{},
) ([]domain.ActionCandidate, error) {
	// Parameters with defaults
	minUnderweightThreshold := GetFloatParam(params, "min_underweight_threshold", 0.05) // 5% underweight
	maxValuePerPosition := GetFloatParam(params, "max_value_per_position", 500.0)
	minScore := GetFloatParam(params, "min_score", 0.6)     // Minimum security score
	maxPositions := GetIntParam(params, "max_positions", 0) // 0 = unlimited

	if !ctx.AllowBuy {
		c.log.Debug().Msg("Buying not allowed, skipping rebalance buys")
		return nil, nil
	}

	if ctx.AvailableCashEUR <= 0 {
		c.log.Debug().Msg("No available cash")
		return nil, nil
	}

	// Check if we have country allocations and weights
	if ctx.CountryAllocations == nil || ctx.CountryWeights == nil {
		c.log.Debug().Msg("No country allocation data available")
		return nil, nil
	}

	c.log.Debug().
		Float64("min_underweight_threshold", minUnderweightThreshold).
		Msg("Calculating rebalance buys")

	// Identify underweight countries
	underweightCountries := make(map[string]float64)
	for country, targetAllocation := range ctx.CountryWeights {
		currentAllocation := ctx.CountryAllocations[country]
		underweight := targetAllocation - currentAllocation
		if underweight > minUnderweightThreshold {
			underweightCountries[country] = underweight
			c.log.Debug().
				Str("country", country).
				Float64("current", currentAllocation).
				Float64("target", targetAllocation).
				Float64("underweight", underweight).
				Msg("Underweight country identified")
		}
	}

	if len(underweightCountries) == 0 {
		c.log.Debug().Msg("No underweight countries")
		return nil, nil
	}

	// Build candidates for securities in underweight countries
	type scoredCandidate struct {
		symbol      string
		group       string
		underweight float64
		score       float64
	}
	var scoredCandidates []scoredCandidate

	for symbol := range ctx.StocksBySymbol {
		// Skip if recently bought
		if ctx.RecentlyBought[symbol] {
			continue
		}

		// Get security and extract country
		security := ctx.StocksBySymbol[symbol]
		country := security.Country
		if country == "" {
			continue // Skip securities without country
		}

		// Map country to group
		group := country
		if ctx.CountryToGroup != nil {
			if mappedGroup, ok := ctx.CountryToGroup[country]; ok {
				group = mappedGroup
			} else {
				group = "OTHER"
			}
		}

		// Check if this group is underweight
		underweight, ok := underweightCountries[group]
		if !ok {
			continue
		}

		// Get security score
		score := 0.5 // Default neutral score
		if ctx.SecurityScores != nil {
			if s, ok := ctx.SecurityScores[symbol]; ok {
				score = s
			}
		}

		// Filter by minimum score
		if score < minScore {
			continue
		}

		scoredCandidates = append(scoredCandidates, scoredCandidate{
			symbol:      symbol,
			group:       group,
			underweight: underweight,
			score:       score,
		})
	}

	// Sort by combined priority (underweight * score)
	sort.Slice(scoredCandidates, func(i, j int) bool {
		priorityI := scoredCandidates[i].underweight * scoredCandidates[i].score
		priorityJ := scoredCandidates[j].underweight * scoredCandidates[j].score
		return priorityI > priorityJ
	})

	// Limit if needed
	if maxPositions > 0 && len(scoredCandidates) > maxPositions {
		scoredCandidates = scoredCandidates[:maxPositions]
	}

	// Create action candidates
	var candidates []domain.ActionCandidate
	for _, scored := range scoredCandidates {
		symbol := scored.symbol
		security := ctx.StocksBySymbol[symbol]

		// Get current price
		currentPrice, ok := ctx.CurrentPrices[symbol]
		if !ok || currentPrice <= 0 {
			c.log.Warn().
				Str("symbol", symbol).
				Msg("No current price available")
			continue
		}

		// Calculate quantity
		targetValue := maxValuePerPosition
		if targetValue > ctx.AvailableCashEUR {
			targetValue = ctx.AvailableCashEUR
		}

		quantity := int(targetValue / currentPrice)
		if quantity == 0 {
			quantity = 1
		}

		// Calculate value
		valueEUR := float64(quantity) * currentPrice

		// Apply transaction costs
		transactionCost := ctx.TransactionCostFixed + (valueEUR * ctx.TransactionCostPercent)
		totalCostEUR := valueEUR + transactionCost

		// Check if we have enough cash
		if totalCostEUR > ctx.AvailableCashEUR {
			continue
		}

		// Priority based on underweight and score
		priority := scored.underweight * scored.score * 0.6

		// Build reason
		reason := fmt.Sprintf("Rebalance: %s underweight by %.1f%% (score: %.2f)",
			scored.group, scored.underweight*100, scored.score)

		// Build tags
		tags := []string{"rebalance", "buy", "underweight"}

		candidate := domain.ActionCandidate{
			Side:     "BUY",
			Symbol:   symbol,
			Name:     security.Name,
			Quantity: quantity,
			Price:    currentPrice,
			ValueEUR: totalCostEUR,
			Currency: string(security.Currency),
			Priority: priority,
			Reason:   reason,
			Tags:     tags,
		}

		candidates = append(candidates, candidate)
	}

	c.log.Info().
		Int("candidates", len(candidates)).
		Int("underweight_countries", len(underweightCountries)).
		Msg("Rebalance buy opportunities identified")

	return candidates, nil
}

func init() {
	// Auto-register on import
	DefaultCalculatorRegistry.Register(NewRebalanceBuysCalculator(zerolog.Nop()))
}
